sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -196,24 +225,19 @@ class FrameLessOffsetWindowFunctionFrame(
     offset: Int,
     ignoreNulls: Boolean = false)
   extends OffsetWindowFunctionFrameBase(
-    target, ordinal, expressions, inputSchema, newMutableProjection, offset) {
+    target, ordinal, expressions, inputSchema, newMutableProjection, offset, ignoreNulls) {
 
-  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
+  override def prepareForDefaultValue(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
+    // During the write phase, always access the current window group. So need reset the states.
     resetStates(rows)
-    if (ignoreNulls) {
-      if (Math.abs(offset) > rows.length) {
-        fillDefaultValue(EmptyRow)
-      } else {
-        findNextRowWithNonNullInput()
-      }
-    } else {
-      // drain the first few rows if offset is larger than zero
-      while (inputIndex < offset) {
-        if (inputIterator.hasNext) inputIterator.next()
-        inputIndex += 1
-      }
-      inputIndex = offset
-    }
+    super.prepareForDefaultValue(rows)
+  }
+
+  override def prepareForRespectNulls(): Unit = {
+    super.prepareForRespectNulls()
+    // In order to match the correct index during the write phase even if the offset is lager than
+    // the window group size or is negative.
```

Comment:
ditto, why can't we always do it? IMO it's more overhead to add these abstractions for such subtle perf difference.